### ISSUE TITLE
configs of training with batch size 256

### ISSUE THREAD

**Issue by poodarchu** [facebookresearch/barlowtwins#7]
can you provide the config to train with batch size 256? how to set the lambd and scale loss? and does the LARS is still necessary when training with small batch sizes?

**Comment by jzbontar** [comment]
> can you provide the config to train with batch size 256? how to set the lambd and scale loss?

We used the following hyperparameters to train our model with a batch size of 256:

```
python main.py /path/to/imagenet/ --epochs 300 --batch-size 256 --learning-rate 0.4 --lambd 0.0051 --projector 8192-8192-8192 --scale-loss 0.024
```

We plan to release a pre-trained model that was trained with a small batch size (256 or 128) shortly. Stay tuned!

**Comment by Mushtaqml** [comment]
If we are training it on a single GPU with batch size of 64, what learning rate we should try with any suggestions? We should be using SGD instead of LARS?

Thanks

**Comment by jzbontar** [comment]
> If we are training it on a single GPU with batch size of 64, what learning rate we should try with any suggestions?

Here is a table of the best learning rate we found for a number of batch sizes (and 300 epochs of training).

| Batch Size | Learning Rate | Top-1 Accuracy on ImageNet |
| --- | --- | --- |
| 128 | 0.7 | 69.7 |
| 256 | 0.4 | 70.7 |
| 512 | 0.3 | 71.4 |
| 1024 | 0.25 | 71.7 |
| 2048 | 0.2 | 71.4 |
| 4096 | 0.2 | 70.7 |

We didn't experiment with a batch size of 64. Judging from the table, a reasonable learning rate for a batch size of 64 would be 1.0.

> We should be using SGD instead of LARS?

It's possible that SGD would work better than LARS for small batch sizes. We did not run that experiment yet.

**Comment by Mushtaqml** [comment]
Thank you @jzbontar for the detailed response, I will test with both the optimizers and share my results. 

Also, I wanted to ask can we try CosineAnnealingLR instead of this adjustment [function](https://github.com/facebookresearch/barlowtwins/blob/52ffe5672cce03761036b401651cdf8459dcb72e/main.py#L146).

**Comment by poodarchu** [comment]
How long does it take to train with batch size 256 on your machine?

**Comment by jzbontar** [comment]
> Thank you @jzbontar for the detailed response, I will test with both the optimizers and share my results.

I'm looking forward to the results.
 
> Also, I wanted to ask can we try CosineAnnealingLR instead of this adjustment [function](https://github.com/facebookresearch/barlowtwins/blob/52ffe5672cce03761036b401651cdf8459dcb72e/main.py#L146).

Yes, go for it. We didn't really experiment too much with the learning rate schedule and our choice of linear warmup followed by cosine decay might not be optimal.

**Comment by jzbontar** [comment]
> How long does it take to train with batch size 256 on your machine? @jzbontar

It depends on the number of GPUs and, consequently, the batch size you use for training. You can use the following table to estimate the training time:

| Batch size per GPU | Time in seconds for 1 epoch on 1 V100 GPU |
| --- | --- |
| 256 | 5284 |
| 128 | 5384 |
| 64 | 5551 |
| 32 | 6608 |
| 16 | 11309 |

For example, the pre-trained model that we made available was trained for 1000 epochs on 16 GPUs with a batch size of 2048 (or a per-gpu batch size of 128).  We can estimate the training time in days as `5384 * 1000 / 16 / 3600 / 24 =
3.9 days`. The actual training time was 7 days, because of the communication overhead between the 16 GPUs.

**Comment by igolan89** [comment]
Hello @jzbontar,
thank you for you amazing work!
I wanted to ask about the batch-size <-> learning-rate numbers. How does this comply with the common scaling rules? Usually when the batch size increases, the learning rate should be increased as well (i.e., [linear scaling rule](https://arxiv.org/pdf/1706.02677.pdf) by Goyal et. al.)

**Comment by jzbontar** [comment]
Hi @igolan89

the numbers in the "Learning Rate" column from the table I posted on Mar 24 are passed to main.py via the --learning-rate flag. These are only *base* learning rates that are multiplied in [main.py](https://github.com/facebookresearch/barlowtwins/blob/c9f549589f0c7c825f7db427a33acc2fc1664149/main.py#L148) by the batch size. So in `main.py` we actually do use the linear scaling rule. For example, the *actual* learning rate in `python main.py --batch-size 32 --learning-rate 0.1` will be twice as large as the learning rate in `python main.py --batch-size 16 --learning-rate 0.1`.

As a side note, the linear scaling rule might not be appropriate for LARS. For example, when using LARS the authors of SimCLR suggest scaling the learning rate by the square root of the batch size.

**Comment by speaker-lover** [comment]
Hello,
what is the scale-loss ?
Best,
